source/dub/recipe/packagerecipe.d

```diff
@@ -171,50 +213,105 @@ struct ConfigurationInfo {
 	}
 }
 
+/// Actual representation of a dependency as specified by the user
+private struct DependencyObject
+{
+    bool optional;
+    @Optional @Name("version")
+    string version_;
+    @Optional string path;
+}
+
+/// Wrapper around `DependencyObject` to do deserialization but still provide
+/// a `Dependency` object to client code.
+private struct YAMLDep
+{
+    private DependencyObject representation;
+    public Dependency actual;
+
+    alias actual this;
+
+    public this (inout(Dependency) dep) inout @safe pure nothrow @nogc
+    {
+        this.actual = dep;
+    }
+
+    public ref YAMLDep opAssign (Dependency dep) return @safe pure nothrow @nogc
+    {
+        this.actual = dep;
+        return this;
+    }
+
+    static YAMLDep fromYAML (scope ConfigParser!YAMLDep p)
+    {
+        import dyaml.node;
+
+        if (p.node.nodeID == NodeID.scalar)
+            return YAMLDep(Dependency(p.node.as!string));
+        auto d = p.parseField!"representation";
+        if (d.path.length)
+        {
+            auto r = YAMLDep(Dependency(NativePath(d.path)));
+            r.optional = d.optional;
+            return r;
+        }
+        else
+        {
+            assert(d.version_.length);
```

Comment:
I don't see why `version_` must have length here, it could be empty if the user didn't specify it or put in an empty string. Use exceptions if you want to actually report configuration issues.

`assert` and function contracts are reserved for testing for programming bugs

Reply:
Actually, this kind of verification should be done in a `validate` method.
Configy will check basic things, like field presence, type, etc... Actual logic (how one field in a struct relates to the others) should be checked by a validate method.